include/lbann/training_algorithms/training_algorithm.hpp

```diff
@@ -0,0 +1,78 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_TRAINING_ALGORITHM_HPP
+#define LBANN_TRAINING_ALGORITHM_HPP
+
+#include "lbann/base.hpp"
+// #include "lbann/comm.hpp"
+// #include "lbann/layers/layer.hpp"
+// #include "lbann/utils/summary.hpp"
+// #include "lbann/utils/graph.hpp"
+// #include "lbann/io/file_io.hpp"
+// #include "lbann/io/persist.hpp"
+// #include "lbann/utils/threads/thread_pool.hpp"
+// #include <lbann.pb.h>
+// #include <vector>
+// #include <string>
+// #include <unordered_map>
+
+#include "lbann/execution_contexts/execution_context.hpp"
+
+namespace lbann {
+
+// Forward-declare this.
+class execution_context;
+class model;
+
+/** Base class for LBANN training_algorithms. */
+class training_algorithm {
+public:
+
+  /** Constructor. */
+  training_algorithm() {};
+  /** Copy constructor. */
+  training_algorithm(const training_algorithm& other) = default;
+  /** Copy assignment operator. */
+  training_algorithm& operator=(const training_algorithm& other) = default;
+  /** Move constructor. */
+  training_algorithm(training_algorithm&& other) = default;
+  /** Move assignment operator. */
+  training_algorithm& operator=(training_algorithm&& other) = default;
+  /** Destructor. */
+  virtual ~training_algorithm() = default;
+  /** Copy training_algorithm. */
+  //  virtual training_algorithm* copy() const = default;
+
+  virtual void apply(execution_context& context,
+                     model& model,
+                     execution_mode mode,
+                     termination_criteria const& term_criteria) = 0;
```

Comment:
If we have a testable use-case for this, I'm in favor of perhaps adding a `vector<model*>&` overload. Otherwise, this should wait until such a use-case exists. Actually, as this PR is already quite large, it might be best to do an "adapting cycleGAN to use new trainer/model/etc interface" PR that implements this.

Reply:
Yes, lets postpone this.